### PR TITLE
most recently used character picker added

### PR DESCRIPTION
**Pull request by 70ray**
This implements a 'most recently used' character picker.
To get the clock icon to work html_safe changed to not double-encode html entities. This should be useful also in tasks.php. Could it cause any problems anywhere else?

### CODE REVIEW

File: tools/proofers/character_selector.js [hunk 1]
```diff
@@ -18,6 +18,84 @@ $(function () {
...
+        });
+        return row;
+    }
+
+    function drawMru() {
+        // this duplicates the html defined in CharacterSelector.inc
+        var mruUpper = mru.slice(0, mruColumns);
+        var mruLower = mru.slice(mruColumns);
+        // this is the selector for mru code
+        $("._26237831463535303b").html(drawRow(mruUpper))
```
Comment (cpeel) [inline tools/proofers/character_selector.js]: What is this ID?

File: pinc/CharacterSelector.inc [hunk 2]
```diff
@@ -6,10 +6,19 @@ class CharacterSelector
     {
         global $utf8_site;
 
+        $mru =
+        [
+            "code" => "&#x1F550;",
```
Comment (cpeel) [inline pinc/CharacterSelector.inc]: Wow, the clock is actually rendered rather large in Chrome on Mac which is unexpected (and, IMHO, unwelcome). That's a little frustrating.

**Comment by cpeel** [comment]
> I want to think about the `html_safe()` change.

After thinking about this and looking at the code, I think your change is the right one.

File: tools/proofers/character_selector.js (hunk 1 above)
Comment (70ray) [reply tools/proofers/character_selector.js]: It was bintohex of the clock entity code. The new change moves all the mru code into the js file so it is not needed now.

**Comment by 70ray** [comment]
This change moves all the mru code into the javascript file.
The change to html_safe is not now needed but it would be useful in another place in putting characters in comments in tasks.php. I could make a new task for it.

File: pinc/CharacterSelector.inc (hunk 2 above)
Comment (70ray) [reply pinc/CharacterSelector.inc]: This is unfortunate.

File: pinc/CharacterSelector.inc [hunk 3]
```diff
@@ -144,7 +144,7 @@ class CharacterSelector
         {
             $code = maybe_utf8_decode($picker["code"]);
             $safe_code = bin2hex($code);
-            $title = maybe_utf8_decode($picker['title']);
+            $title = attr_safe(maybe_utf8_decode($picker['title']));
             $selector_string .= "<button type='button' id='_$safe_code' class='selector_button' title='$title'>" . html_safe($code) . "</button>";
```
[the remark this answers is not in the document]
Comment (70ray) [reply pinc/CharacterSelector.inc]: I removed the ID so it's not needed.

File: tools/proofers/ctrl_frame.php [hunk 4]
```diff
@@ -5,15 +5,23 @@
 include_once($relPath.'stages.inc');
 include_once($relPath.'misc.inc'); // get_enumerated_param()
 include_once($relPath.'ProofreadingToolbox.inc');
+include_once($relPath.'Project.inc'); // validate_projectID()
+
 
 $round_id = get_enumerated_param($_GET, 'round_id', null, array_keys($Round_for_round_id_));
 $round = get_Round_for_round_id($round_id);
+$projectid = validate_projectID('projectid', @$_GET['projectid']);
+$mru_title = _("Most recently used");
```
[the remark this answers is not in the document]
Comment (70ray) [reply tools/proofers/ctrl_frame.php]: Yes we do.

File: pinc/CharacterSelector.inc (hunk 3 above)
[the remark this answers is not in the document]
Comment (cpeel) [reply pinc/CharacterSelector.inc]: I don't understand your response. If a developer changes the HTML for how the buttons are rendered (changing them to `<div>`s or something) they need to update two places -- here and in the .js file. My suggestion was to add a comment here letting them know we need to keep the two in sync.

[the remark this answers is not in the document]
Comment (70ray) [reply pinc/CharacterSelector.inc]: I see what you mean, I was still thinking about the connection using the strange bin_to_hex code.

**Comment by 70ray** [comment]
@cpeel It seems I am not able to request reviews myself but please feel free do do so.

**Comment by cpeel** [comment]
Code-wise this looks great. I'd like for us to have some folks in the community test it out and give feedback.

**Comment by 70ray** [comment]
I've made a post in the future features forum.

**Pull request merged**